### PR TITLE
Apply “role” attributes for better accessibility

### PR DESCRIPTION
**Pull request by tedw**
Regarding #124, I took a stab at adding `role` attributes to the `flickity-slider` wrapper and individual cells so that assistive devices will treat them as a list.

### CODE REVIEW

**Comment by tedw** [comment]
I think we should also hide the dots list from screen readers unless you want to go to the trouble of making them function like tabs, like in this example http://jongund.github.io/aria-examples/bootstrap-carousel/carousel-3.html

**Comment by tedw** [comment]
So it seems adding `tabindex="0"` to the wrapper div causes VoiceOver to read _all_ of the inner content at once 😕  Perhaps enabling the left/right arrow key navigation only when the dots or nav buttons are focused would allow us to remove `tabindex`. The dots would need to be updated to work like accessible tabs, though ([example](http://heydonworks.com/practical_aria_examples/#tab-interface)).

**Comment by desandro** [comment]
Hi Ted, Thanks so much for your work here. I'll follow up with comments

File: js/flickity.js [hunk 1]
```diff
@@ -232,6 +238,11 @@ proto.reloadCells = function() {
 proto._makeCells = function( elems ) {
   var cellElems = this._filterFindCellElements( elems );
 
+  // Add role="listitem" to cells for improved accessibility
+  for ( var i=0; i < cellElems.length; i++ ) {
+    cellElems[i].setAttribute('role', 'listitem');
+  }
+
```
Comment (desandro) [inline js/flickity.js]: This code should happen in cell.js

File: js/flickity.js [hunk 2]
```diff
@@ -179,8 +185,6 @@ proto.activate = function() {
   this.reloadCells();
 
   if ( this.options.accessibility ) {
-    // allow element to focusable
-    this.element.tabIndex = 0;
```
Comment (desandro) [inline js/flickity.js]: So what gets to be tab-able if the main element is not?

File: js/flickity.js [hunk 3]
```diff
@@ -764,22 +775,25 @@ proto.watchCSS = function() {
...
-    var leftMethod = this.options.rightToLeft ? 'next' : 'previous';
-    this.uiChange();
-    this[ leftMethod ]();
-  } else if ( event.keyCode == 39 ) {
-    // go right
-    var rightMethod = this.options.rightToLeft ? 'previous' : 'next';
-    this.uiChange();
-    this[ rightMethod ]();
+  // only work if child of carousel is focused
+  if ( this.options.accessibility && this.element.contains(document.activeElement) ) {
```
Comment (desandro) [inline js/flickity.js]: I like how this works by looking for the focused element within the Flickity container, but I'm worried that its to general. For example, you could have a text input that needs to have separate keyboard controls from Flickity's behavior.

File: js/page-dots.js [hunk 4]
```diff
@@ -46,6 +46,8 @@ PageDots.prototype._create = function() {
   // create holder element
   this.holder = document.createElement('ol');
   this.holder.className = 'flickity-page-dots';
+  // hide from screen readers
+  this.holder.setAttribute('aria-hidden', 'true');
```
Comment (desandro) [inline js/page-dots.js]: Ah, thanks for this.

**Comment by tedw** [comment]
@desandro Thanks for feedback! Sorry for delay, haven’t had much time to revisit this (have a 2 month-old at home).

There is some good keyboard nav guidance here:
https://www.w3.org/TR/wai-aria-practices-1.1/#kbd_general_within

(via [Roving tabindex -- A11ycasts #06](https://youtu.be/uCIC2LNt0bk))

I now think your original approach, making the whole container focusable, is best, combined with the “roving tabindex” technique described above.

Additionally, it might make sense to use the following:

`role="application"`
https://www.w3.org/TR/wai-aria-1.1/#application

`aria-activedescendant`
https://www.w3.org/TR/wai-aria-practices-1.1/#focus_activedescendant

I also found this guide https://www.w3.org/WAI/tutorials/carousels/ but it doesn't seem to be as screen reader friendly as this one http://jongund.github.io/aria-examples/bootstrap-carousel/carousel-3.html I mentioned before.

**Comment by tedw** [comment]
Another take on carousels from Estelle Weyl, uses radio buttons and labels:
https://github.com/estelle/merry-go-round

**Comment by stevepiron** [comment]
Hello, any chance of getting that PR merged soon? That's a very nice plus, imo :)

**Comment by desandro** [comment]
This PR has grown a bit old. I'm now going to close.

But, I'm still open to implementing the feature — specifically adding a `role` attribute. For other related accessibility issues see #776 and #777

**Comment by iosonogio** [comment]
Hi, this improvement (about adding "role = list") does not seem to be there in code version 2.2.1

**Comment by vasanthangel4** [comment]
Hi @desandro :  when we enabling the voice over and accessibility carousel scrolling not working with Three-finger triple tap in  Safari browser in all ios device. android application is working fine.  when we disablng the accessibility carousel scoling/swipe is working fine. please let us know if you have solution. i have applied accessibility improvement PR, but it is not helping. please let me know if any vanila js fix for this issue.

**Pull request closed without merging**